safekeeper/src/wal_backup.rs

```diff
@@ -532,6 +539,35 @@ pub async fn read_object(
     Ok(Box::pin(reader))
 }
 
+/// Delete WAL files for the given timeline. Remote storage must be configured
+/// when called.
+pub async fn delete_timeline(ttid: &TenantTimelineId) -> Result<()> {
+    let storage = get_configured_remote_storage();
+    let ttid_path = Utf8Path::new(&ttid.tenant_id.to_string()).join(ttid.timeline_id.to_string());
+    let remote_path = RemotePath::new(&ttid_path)?;
+
+    // A backoff::retry is used here for two reasons:
+    // - To provide a backoff rather than busy-polling the API on errors
+    // - To absorb transient 429/503 conditions without hitting our error
+    //   logging path for issues deleting objects.
+    let token = CancellationToken::new(); // not really used
+    backoff::retry(
+        || async {
+            let files = storage.list_files(Some(&remote_path)).await?;
```

Comment:
There can be issues with it for hugely large projects. HTTP endpoint for deletion can be cancelled on timeout, and if timeout is lower than time it takes to list all files, we will be stuck a loop without a chance to finish deletion.

In other words, there will be no progress if future will always be cancelled while in `list_files`, without reaching to `delete_objects`.

Probably the best way to proceed here is to create an issue and add a comment here, because it should be very rare and the fix is not trivial.

Reply:
the file listing call also might not return all files but only a subset as part of a segmented response. But this is also a thing to worry about in the future.

Reply:
> the file listing call also might not return all files but only a subset as part of a segmented response. But this is also a thing to worry about in the future.

s3_bucket.rs has support for returning everything, see continuation_token there.

Reply:
> HTTP endpoint for deletion can be cancelled on timeout, and if timeout is lower than time it takes to list all files, we will be stuck a loop without a chance to finish deletion.

I don't think we have http requests timeouts in safekeepers. But we probably should, so I'll create an issue and add comment.

Reply:
added comment

Reply:
> I don't think we have http requests timeouts in safekeepers. But we probably should, so I'll create an issue and add comment.

I'm talking about control-plane -> safekeeper HTTP requests. We have timeouts in there.

Reply:
Right, but I suspect that the call on sk will continue execution regardless of fallen off caller (cplane). Not absolutely sure, depends on hyper, but likely it would.

Reply:
I remember it is being cancelled, but can't say 100%. Need to check.